engines/e_pkcs11.c

```diff
@@ -0,0 +1,795 @@
+static int pkcs11_parse_uri(const char *path, char *token, char **value);
+static int pkcs11_parse(PKCS11_CTX *ctx, const char *path);
+static EVP_PKEY *pkcs11_load_pkey(PKCS11_CTX *ctx);
+static char pkcs11_hex_int(char nib1, char nib2);
+static int pkcs11_rsa_enc(int flen, const unsigned char *from,
+                          unsigned char *to, RSA *rsa, int padding);
+static RSA_METHOD *pkcs11_rsa(void);
+static PKCS11_CTX *pkcs11_ctx_new(void);
+static void pkcs11_ctx_free(PKCS11_CTX *ctx);
+static CK_SLOT_ID pkcs11_get_slot(PKCS11_CTX *ctx);
+static CK_OBJECT_HANDLE pkcs11_get_private_key(PKCS11_CTX *ctx);
+static CK_FUNCTION_LIST *pkcs11_funcs;
+static void PKCS11_trace(char *format, ...);
+
+/* Engine stuff */
+static int pkcs11_init(ENGINE *e);
+static int pkcs11_bind(ENGINE *e, const char *id);
+static int pkcs11_destroy(ENGINE *e);
+static int pkcs11_finish(ENGINE *e);
+static EVP_PKEY *pkcs11_engine_load_private_key(ENGINE * e, const char *path,
+                                                UI_METHOD * ui_method,
+                                                void *callback_data);
+static const char *engine_id = "pkcs11";
+static const char *engine_name = "A minimal PKCS#11 engine only for sign";
+static int pkcs11_idx = -1;
+typedef CK_RV pkcs11_pFunc(CK_FUNCTION_LIST **pkcs11_funcs);
+
+static int pkcs11_init(ENGINE *e)
+{
+    PKCS11_CTX *ctx;
+
+    if (pkcs11_idx < 0) {
+        pkcs11_idx = ENGINE_get_ex_new_index(0, NULL, NULL, NULL, 0);
+        if (pkcs11_idx < 0)
+            goto memerr;
+    }
+    ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+    if (ctx == NULL) {
+        ctx = pkcs11_ctx_new();
+
+        if (ctx == NULL)
+            goto memerr;
+
+        ENGINE_set_ex_data(e, pkcs11_idx, ctx);
+    }
+    return 1;
+
+ memerr:
+    PKCS11err(PKCS11_F_PKCS11_INIT, ERR_R_MALLOC_FAILURE);
+    return 0;
+}
+
+static int pkcs11_ctrl(ENGINE *e, int cmd, long i, void *p, void (*f) (void))
+{
+    int ret = 1;
+    char *tmpstr;
+
+    PKCS11_CTX *ctx;
+
+    if (pkcs11_idx == -1 && !pkcs11_init(e)) {
+        PKCS11err(PKCS11_F_PKCS11_CTRL, PKCS11_R_ENGINE_NOT_INITIALIZED);
+        return 0;
+    }
+    ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+
+    switch (cmd) {
+    case PKCS11_CMD_MODULE_PATH:
+        tmpstr = OPENSSL_strdup(p);
+        if (tmpstr != NULL) {
+            ctx->module_path = tmpstr;
+            PKCS11_trace("Setting module path to %s\n", ctx->module_path);
+        } else {
+            PKCS11err(PKCS11_F_PKCS11_CTRL, ERR_R_MALLOC_FAILURE);
+            ret = 0;
+        }
+        break;
+
+    case PKCS11_CMD_PIN:
+        tmpstr = OPENSSL_strdup(p);
+        if (tmpstr != NULL) {
+            ctx->pin = (CK_BYTE *) tmpstr;
+            PKCS11_trace("Setting pin\n");
+        } else {
+            PKCS11err(PKCS11_F_PKCS11_CTRL, ERR_R_MALLOC_FAILURE);
+            ret = 0;
+        }
+        break;
+    }
+
+    if (ret)
+        ENGINE_set_ex_data(e, pkcs11_idx, ctx);
```

Comment:
AS your method don't reallocate `ctx`, there no need to call `ENGINE_set_ex_data` again.